src/python/pants/backend/python/tasks/select_interpreter.py

```diff
@@ -24,16 +25,19 @@
 class PythonInterpreterFingerprintStrategy(DefaultFingerprintHashingMixin, FingerprintStrategy):
 
   def compute_fingerprint(self, python_target):
-    # Only consider the compatibility requirements in the fingerprint, as only
+    # Consider the compatibility requirements from both the targets and global constraints, as only
     # those can affect the selected interpreter.
+    interpreter_constraints = PythonSetup.global_instance().interpreter_constraints
```

Comment:
This line is probably not quite right. A couple issues I see:
1) Would we need to add `PythonSetup` to `SelectInterpreter`'s `subsystem_dependencies`?
2) Is this duplicating logic? `PythonInterpreterCache` already has a property `_python_setup`. We could reference it through the following diff:
```diff
diff --git a/src/python/pants/backend/python/tasks/select_interpreter.py b/src/python/pants/backend/python/tasks/select_interpreter.py
index dbc31b1e3..f0cdebe84 100644
--- a/src/python/pants/backend/python/tasks/select_interpreter.py
+++ b/src/python/pants/backend/python/tasks/select_interpreter.py
@@ -13,7 +13,6 @@ from pex.executor import Executor
 from pex.interpreter import PythonInterpreter

 from pants.backend.python.interpreter_cache import PythonInterpreterCache
-from pants.backend.python.subsystems.python_setup import PythonSetup
 from pants.backend.python.targets.python_requirement_library import PythonRequirementLibrary
 from pants.backend.python.targets.python_target import PythonTarget
 from pants.base.fingerprint_strategy import DefaultFingerprintHashingMixin, FingerprintStrategy
@@ -24,10 +23,13 @@ from pants.util.dirutil import safe_mkdir_for

 class PythonInterpreterFingerprintStrategy(DefaultFingerprintHashingMixin, FingerprintStrategy):

+  def __init__(self, interpreter_cache):
+    self.interpreter_cache = interpreter_cache
+
   def compute_fingerprint(self, python_target):
     # Consider the compatibility requirements from both the targets and global constraints, as only
     # those can affect the selected interpreter.
-    interpreter_constraints = PythonSetup.global_instance().interpreter_constraints
+    interpreter_constraints = self.interpreter_cache._python_setup.interpreter_constraints
     hash_elements_for_target = []
     if python_target.compatibility:
       hash_elements_for_target.extend(sorted(python_target.compatibility))
@@ -69,7 +71,7 @@ class SelectInterpreter(Task):
     if not python_tgts_and_reqs:
       return
     python_tgts = [tgt for tgt in python_tgts_and_reqs if isinstance(tgt, PythonTarget)]
-    fs = PythonInterpreterFingerprintStrategy()
+    fs = PythonInterpreterFingerprintStrategy(interpreter_cache=PythonInterpreterCache.global_instance())
     with self.invalidated(python_tgts, fingerprint_strategy=fs) as invalidation_check:
       # If there are no relevant targets, we still go through the motions of selecting
       # an interpreter, to prevent downstream tasks from having to check for this special case.
```
3. Is it wrong to call `global_instance()`? I'm a bit confused how that impacts testing.

Reply:
Invalidation already mixes in the task's own fingerprint, which includes the fingerprints of all subsystems in its closure. See https://github.com/pantsbuild/pants/blob/7819724acb0f0e6fd83a3cd9ec6b5b72a9799721/src/python/pants/task/task.py#L310.

So I think all you need to do is to add `fingerprint=True` on the registration of `--interpreter-constraints` in `PythonSetup`.  You shouldn't need to change anything here. 

As a general comment: If the above were not true, any information you wanted to use to similar effect would need to come from `PythonInterpreterCache.global_instance()`, because that's what this task actually uses in order to select an interpreter. The fact that `PythonInterpreterCache` happens to use `PythonSetup.global_instance()` is not something that this task should have to know about. In fact, we should remove `PythonSetup` from this task's `subsystem_dependencies`, since it doesn't use it directly.